src/main/java/graphql/nadel/hooks/ServiceExecutionHooks.java

```diff
@@ -1,55 +1,48 @@
+import graphql.PublicSpi;
 import graphql.execution.nextgen.result.RootExecutionResultNode;
-import graphql.nadel.Service;
-import graphql.schema.GraphQLSchema;
 
 import java.util.concurrent.CompletableFuture;
 
 /**
  * These hooks allow you to change the way service execution happens
  */
+@PublicSpi
 public interface ServiceExecutionHooks {
 
 
     /**
-     * Called per top level field that for a service.  This allows you to create a "context" object that will be passed into further calls.
+     * Called per top level field for a service.  This allows you to create a "context" object that will be passed into further calls.
      *
-     * @param service          the service in play
-     * @param topLevelStepInfo the top level field execution step info
+     * @param params the parameters to this call
      *
      * @return an async context object of your choosing
      */
-    default CompletableFuture<Object> createServiceContext(Service service, ExecutionStepInfo topLevelStepInfo) {
+    default CompletableFuture<Object> createServiceContext(CreateServiceContextParams params) {
         return CompletableFuture.completedFuture(null);
     }
 
     /**
      * Called to possibly change the arguments and runtime variables that are passed onto a called service
      *
-     * @param service          the service in play
-     * @param serviceContext   the context object created in {@link #createServiceContext(graphql.nadel.Service, graphql.execution.ExecutionStepInfo)}
-     * @param topLevelStepInfo the top level field execution step info
+     * @param params the parameters to this call
      *
-     * @return an async null to indicate NO change needed or an async new set of ModifiedArguments
+     * @return an async null to indicate NO change needed or an async new document and variables
      */
-    default CompletableFuture<ModifiedArguments> modifyArguments(Service service, Object serviceContext, ExecutionStepInfo topLevelStepInfo) {
+    default CompletableFuture<QueryRewriteResult> queryRewrite(QueryRewriteParams params) {
         return CompletableFuture.completedFuture(null);
     }
 
     /**
      * Called to allow a service to post process the service result in some fashion.
      *
-     * @param service        the service in play
-     * @param serviceContext the context object created in {@link #createServiceContext(graphql.nadel.Service, graphql.execution.ExecutionStepInfo)}
-     * @param overallSchema  the overall schema
-     * @param resultNode     the result
+     * @param params the parameters to this call
      *
      * @return an async possible result node
      */
-    default CompletableFuture<RootExecutionResultNode> postServiceResult(Service service, Object serviceContext, GraphQLSchema overallSchema, RootExecutionResultNode resultNode) {
-        return CompletableFuture.completedFuture(resultNode);
+    default CompletableFuture<RootExecutionResultNode> resultRewrite(ResultRewriteParams params) {
```

Comment:
this is a bit inconsistent : one method returns null, one the input

Reply:
I wanted to have a short cut - that is a signal that NOTHING needs to change at all.  

I could have made it return Optional I guess